src/mbgl/text/glyph_set.cpp

```diff
@@ -81,88 +80,168 @@ void align(Shaping& shaping,
+    while (priorBreak) {
+        leastBadBreaks.insert(priorBreak->index);
+        priorBreak = priorBreak->priorBreak.get();
+    }
+    return leastBadBreaks;
+}
+
 
 // We determine line breaks based on shaped text in logical order. Working in visual order would be
 //  more intuitive, but we can't do that because the visual order may be changed by line breaks!
-std::set<int32_t> GlyphSet::determineLineBreaks(const std::u16string& logicalInput,
+std::set<std::size_t> GlyphSet::determineLineBreaks(const std::u16string& logicalInput,
                                                 const float spacing,
                                                 float maxWidth) const {
-    if (!maxWidth)
+    if (!maxWidth) {
         return {};
+    }
 
-    if (logicalInput.empty())
+    if (logicalInput.empty()) {
         return {};
-
-    if (util::i18n::allowsIdeographicBreaking(logicalInput))
-        maxWidth = determineIdeographicLineWidth(logicalInput, spacing, maxWidth);
-
-    std::set<int32_t> lineBreakPoints;
+    }
+    
+    const float targetWidth = determineAverageLineWidth(logicalInput, spacing, maxWidth);
+    
+    std::vector<std::shared_ptr<PotentialBreak>> potentialBreaks;
     float currentX = 0;
-    uint32_t lastSafeBreak = 0;
-    float lastSafeBreakX = 0;
 
-    for (uint32_t i = 0; i < logicalInput.size(); i++) {
-        auto it = sdfs.find(logicalInput[i]);
-        if (it == sdfs.end())
+    for (std::size_t i = 0; i < logicalInput.size(); i++) {
+        const char16_t codePoint = logicalInput[i];
+        auto it = sdfs.find(codePoint);
+        if (it == sdfs.end()) {
             continue;
-
-        const SDFGlyph& glyph = it->second;
+        }
 
         // Ideographic characters, spaces, and word-breaking punctuation that often appear without
         // surrounding spaces.
-        if (util::i18n::allowsWordBreaking(glyph.id) ||
-            util::i18n::allowsIdeographicBreaking(glyph.id)) {
-            lastSafeBreak = i;
-            lastSafeBreakX = currentX;
-        }
+        if (util::i18n::allowsWordBreaking(codePoint) ||
+            util::i18n::allowsIdeographicBreaking(codePoint)) {
+            const char16_t previousCodePoint = i > 0 ? logicalInput[i-1] : 0;
 
-        if (currentX > maxWidth && lastSafeBreak > 0) {
-            lineBreakPoints.insert(lastSafeBreak);
-            currentX -= lastSafeBreakX;
-            lastSafeBreakX = 0;
+            potentialBreaks.push_back(evaluateBreak(i, currentX, targetWidth, potentialBreaks,
+                                                     calculatePenalty(codePoint, previousCodePoint),
+                                                     false));
         }
 
-        currentX += glyph.metrics.advance + spacing;
+        currentX += it->second.metrics.advance + spacing;
     }
+    potentialBreaks.push_back(evaluateBreak(logicalInput.size(), currentX, targetWidth, potentialBreaks, 0, true));
```

Comment:
Return the result directly rather than doing `push_back` + `back`: `return leastBadBreaks(*evaluateBreak(...));`